### DISCUSSION TITLE
Incorrect anchor handling

### DISCUSSION THREAD

**Discussion by daniser** [tighten/ziggy#638] [General]
### Ziggy version

v1.4.6

### Laravel version

v9.8.1

### Description

Laravel's URL generator have separate algorithm for handling URL fragments (anchors?):
`\Illuminate\Routing\RouteUrlGenerator::addQueryString`.
Fragments are always appended to the end of query string and have no '=' after them.

Ziggy lacks that handy feature. That's the difference between Laravel and Ziggy generated URLs for the same route:

Laravel/Blade:
```php
route('order', [123, '#details', 'param' => 'test']);
```
=>`https://domain.com/orders/123?param=test&#details`

Ziggy:
```js
route('order', [123, '#details', {param: 'test'}]);
```
=>`https://domain.com/orders/123?#details=&param=test`

### Ziggy call and context

```js
route('order', [123, '#details', {param: 'test'}]);
```

### Ziggy configuration

```js
{
    "url": "https://domain.com",
    "port": null,
    "defaults": {},
    "routes": {
        "order": {
            "uri": "orders/{order}",
            "methods": [
                "GET",
                "HEAD"
            ],
            "wheres": {
                "order": "\\d+"
            }
        }
    }
}
```

### Route definition

```php
Route::get('/orders/{order}', [OrderController::class, 'index'])->where('order', '\d+')->name('order');
```

**Comment by bakerkretzmar** [comment]
I dug into this some more and it's actually not the case that Laravel automatically sorts params starting with `#` to the end of the URL. It puts _numeric-indexed_ params last, and in your example that happens to produce the same result.

Here's another example with more params:

```php
Route::get('/orders/{order}', [OrderController::class, 'index'])->name('order');

route('order', [123, 'first', '#anchor', 'third', 'param' => 'test']);

// https://domain.com/orders/123?param=test&first&#anchor&third
```

String-indexed params, in this case `'params' => 'test'`, are formatted as query parameters and output first, and then other params are just added on as query params in the order they were provided.

What [`RouteUrlGenerator::addQueryString`](https://github.com/laravel/framework/blob/1c783fea3f845d4bb886e9e33cab0c7ed0007812/src/Illuminate/Routing/RouteUrlGenerator.php#L239) is doing is ensuring that if the URL _already_ has a fragment it's _kept_ at the end. So this works, which is neat:

```php
Route::get('orders/{order}#shipping', [OrderController::class, 'index'])->name('order-shipping');

route('order-shipping', [123, 'foo', 'param' => 'test']);

// https://domain.com/orders/123?param=test&foo#shipping
```

I'm not sure there's actually anything Ziggy should do differently here.

**Comment by mrleblanc101** [comment]
In vue, I had to do this ``:href="`${route('terms-and-conditions')}#no-show`"`` to add an anchor since I couldn't find a way to add an anchor any other way.
As OP said, `route('terms-and-conditions', ['#no-show']);` results in `/terms-and-conditions#no-show=` which is not correct since `#no-show` is an anchor and not a param
